src/Illuminate/Database/Eloquent/Builder.php

```diff
@@ -556,6 +557,22 @@ public function firstOrCreate(array $attributes = [], array $values = [])
         });
     }
 
+    /**
+     * Concurrent-safe load first record matching the attributes or create it.
+     *
+     * @param  array  $attributes
+     * @param  array  $values
+     * @return \Illuminate\Database\Eloquent\Model|static
+     */
+    public function concurrentlyFirstOrCreate(array $attributes = [], array $values = [])
+    {
+        try {
+            return $this->firstOrCreate($attributes, $values);
+        } catch (QueryException $e) {
```

Comment:
The problem with catching `QueryException` is that we'll end up trying to re-do something that should have failed.

For example;

```php
User::firstOrCreate(['id' => 11]);
```

In the Laravel skeleton, the `name` field is also required, so now `firstOrCreate` will throw `QueryException` and we'll try again to find a matching model.

Reply:
> https://stackoverflow.com/questions/27878719/laravel-catch-eloquent-unique-field-error

This will change per database engine.